### DISCUSSION TITLE
DAGs from external PVC (Azure blob storage) are created, but never updated until pod recycle

### DISCUSSION THREAD

**Discussion by mblumenshine** [apache/airflow#33928] [Q&A] [answered]
Hi everyone,

I want to confirm if this is the intended use or not for external PVCs. We are extending the official docker image, using Airflow 2.7.0, Python 3.11, and the official helm chart. KubernetesExecutor, running in AKS. Airflow works fine except that DAGs do not update, both in the UI and that which is executed (interestingly, existing tasks will run the latest code, but any new tasks that are added will not be run).

We are managing our DAG files with Azure blob storage. Users either use a pipeline that copies their DAGs into blob storage, or often they use Azure Portal directly. If we recycle the pods, we see `/opt/airflow/dags` will contain the latest DAGs. When we don't recycle the pods and just make changes to blob storage, if we add new DAGs, those will appear in `/opt/airflow/dags`. However, existing DAGs never update until a pod recycle.

Are we interpreting the doc correctly? Are we supposed to somehow update the container and simply use DAG persistence for upgrades/restarts/downtime? Or should we be able to push a file to blob storage and expect it to update? I'm relatively new to Airflow, but the folks who were working on this previously seem convinced that DAGs edited in Azure would be reflected in Airflow within minutes.

There are no errors in the scheduler or webserver logs.

Thank you!

**Comment by mblumenshine** [comment]
So, it sounds like Airflow is not responsible for syncing up blob storage and the PVC and that we'd need to ensure the PVC is updated. That sound right?

**Reply by potiuk** [reply, marked as answer]
Files need to find its way to the /opt/airflow/dags - in the mounted container. You have not described how the blob-storate and PVC gets synchronized it. Generally when the dags are presente in the PVC in the containers, Airflow scheduler will scan the folder regularly (see configuration of scheduler) and adds any new files it finds there. But alrflow has no knowledge about the blob-storate of yours - if you have some syncing between the PVC where the dags are and teh blob store - you need to run it when the files are updated in the blob-store. But this is a question of how your deployment and synchronization is done. Airflow just perferms scanning of the /opt/airflow/dags - and files must be there for it to find it.

**Reply by mblumenshine** [reply]
Excellent, thank you! I suspected as much. I believe this was a big misunderstanding, was never finished, and yet folks had the impression it was "working" because the pods were recycling over and over.

Thanks a million!

**Reply by escortnotice** [reply]
@mblumenshine  -- how did you solve your issue finally. I am also facing the same issue you had posted. Do you have a mail id I can get in touch and try to understand how you have solved your issue.

**Reply by mblumenshine** [reply]
We decided to replace blob storage with git sync. We had our DAGs scattered in a number of different git repositories, so we created a new repo for the purpose of git sync and used git submodules to connect them all.

If you're using blob storage, blob storage is only meant for **persistence**. It's a store for DAGs when the server restarts and DAGs need to be loaded in. If your workflow for adding and updating DAGs is by uploading them to blob storage, you'll need some code that responds to those changes and programmatically pulls your DAGs from blob storage and brings them into the Airflow containers at /opt/airflow/dags.

If you update DAGs infrequently, perhaps just restart your Airflow cluster when you update your DAGs in blob storage.

If you update your DAGs frequently, I'd recommend just using git syncing.

**Reply by escortnotice** [reply]
Thank You so much for your response. I think I understand now. Will try the steps you mentioned.

**Reply by escortnotice** [reply]
@mblumenshine , I was finally able to solve the sync issue using Azure Blob as the storage for the dag bag. Below is the solution which worked for me.

The issue is that this data is cached at two levels as below: 
1)	The virtual file system driver for Azure Blob Storage caches the content
2)	The kernel page caches the content


Use the below parameters while creating the PV which will override the default caching behaviour. 

  mountOptions:
    - '-o allow_other'
    - '--file-cache-timeout-in-seconds=1'
    - '-o direct_io'

**Reply by edgBR** [reply]
Hi @escortnotice and @potiuk 

I am having similar issue here:

https://github.com/apache/airflow/discussions/54937

But unfortunately I am still a bit confused as in my case changing the dag name makes the sync work properly.
